File: Chapter06/Renderer.cpp
//
//  Renderer.cpp
//  Game-mac
//
//  Created by Michael Lehmann on 14.01.21.
//  Copyright © 2021 Sanjay Madhav. All rights reserved.
//

#include "Renderer.h"
#include "Texture.h"
#include "Mesh.h"
#include <algorithm>
#include "Shader.h"
#include "VertexArray.h"
#include "SpriteComponent.h"
#include "MeshComponent.h"
#include <GL/glew.h>

Renderer::Renderer(Game* game)
    :mGame(game)
    ,mSpriteShader(nullptr)
    ,mMeshShader(nullptr)
{
}

Renderer::~Renderer()
{
}

bool Renderer::Initialize(float screenWidth, float screenHeight)
{
    mScreenWidth = screenWidth;
    mScreenHeight = screenHeight;

    // Set OpenGL window's attribute (use prior to creating the window)
    // Returns 0 if successful, othewise a negative value
    // use the core OpenGL profile
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);
    
    // Specify version 3.3
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 3);
    
    // Request a color buffer with 8-bits per RGBA channel
    SDL_GL_SetAttribute(SDL_GL_RED_SIZE, 8);
    SDL_GL_SetAttribute(SDL_GL_GREEN_SIZE, 8);
    SDL_GL_SetAttribute(SDL_GL_BLUE_SIZE, 8);
    SDL_GL_SetAttribute(SDL_GL_ALPHA_SIZE, 8);
    // SDL_GL_SetAttribute(SDL_GL_DEPTH_SIZE, 24);
    
    // Enable double buffering
    SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, 1);
    
    // Force OpenGL to use hardware acceleration
    SDL_GL_SetAttribute(SDL_GL_ACCELERATED_VISUAL, 1);

    // Create an SDL Window
    // See: https://discourse.libsdl.org/t/sdl2-cant-see-the-window/25835
    mWindow = SDL_CreateWindow(
            "Game Programming in C++ (Chapter 6)", // Window title
            100,    // Top left x-coordinate of window
            100,    // Top left y-coordinate of window
            static_cast<int>(mScreenWidth),   // Width of window
            static_cast<int>(mScreenHeight),    // Height of window
            SDL_WINDOW_OPENGL);   // Flags (0 for no flags set)

    if (!mWindow)
    {
        SDL_Log("Failed to create window: %s", SDL_GetError());
        return false;
    }
    
    // Create an OpenGL context
    mContext = SDL_GL_CreateContext(mWindow);
    
    // Initialize GLEW
    glewExperimental = GL_TRUE;
    if (glewInit() != GLEW_OK)
    {
        SDL_Log("Failed to initialize GLEW.");
        return false;
    }
    
    // On some platforms, GLEW will emit a benign error code,
    // so clear it
    glGetError();
    
    // Make sure we can create/compile shaders
    if (!LoadShaders())
    {
        SDL_Log("Failed to load shaders.");
        return false;
    }
    
    // Create quad for drawing sprites
    CreateSpriteVerts();

    return true;
}

void Renderer::Shutdown()
{
    delete mSpriteVerts;
    mSpriteShader->Unload();
    delete mSpriteShader;
    mMeshShader->Unload();
    delete mMeshShader;
    SDL_GL_DeleteContext(mContext);
    SDL_DestroyWindow(mWindow);
}

void Renderer::UnloadData()
{
    // Destroy textures
    for (auto i : mTextures)
    {
        i.second->Unload();
        delete i.second;
    }
    
    mTextures.clear();
    
    // Destroy meshes
    for (auto i : mMeshes)
    {
        i.second->Unload();
        delete i.second;
    }
    
    mMeshes.clear();
}

void Renderer::Draw()
{
    // Set the clear color to light gray
    glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
    
    // Clear the color buffer
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
    
    // Draw all mesh components
    // Enable depth buffer/disable alpha blend
    glEnable(GL_DEPTH_TEST);
    glDisable(GL_BLEND);
    
    // Set the basic mesh shader active
    mMeshShader->SetActive();
    
    // Update view-projection matrix
    mMeshShader->SetMatrixUniform("uViewProj", mView * mProjection);
    
    // Update lighting uniforms
    SetLightUniforms(mMeshShader);
    
    // Draw all meshs
    for (auto mesh : mMeshComps)
    {
        mesh->Draw(mMeshShader);
    }
    
    // Draw all sprite components
    // Disable depth buffering
    glDisable(GL_DEPTH_TEST);
    glEnable(GL_BLEND);
    glBlendEquationSeparate(GL_FUNC_ADD, GL_FUNC_ADD);
    glBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA, GL_ONE, GL_ZERO);
    
    // Set sprite shader and vertex array objects active
    mSpriteShader->SetActive();
    mSpriteVerts->SetActive();
    
    // Draw all sprites
    for (auto sprite : mSprites)
    {
        sprite->Draw(mSpriteShader);
    }
    
    // Swap the buffers, which also displays the scene
    SDL_GL_SwapWindow(mWindow);
}

void Renderer::AddSprite(class SpriteComponent *sprite) {
    // Find the insertion point in the sorted vector
    // (The first element with a higher draw order than me)
    int myDrawOrder = sprite->GetDrawOrder();
    auto iter = mSprites.begin();
    for( ; iter != mSprites.end(); ++iter)
    {
        if (myDrawOrder < (*iter)->GetDrawOrder())
        {
            break;
        }
    }
    
    // inserts element before position iterator
    mSprites.insert(iter, sprite);
}

void Renderer::RemoveSprite(SpriteComponent* sprite)
{
    // (We can't swap because it ruins ordering)
    auto iter = std::find(mSprites.begin(), mSprites.end(), sprite);
    mSprites.erase(iter);
}

void Renderer::AddMeshComp(MeshComponent* mesh)
{
    mMeshComps.emplace_back(mesh);
}

void Renderer::RemoveMeshComp(MeshComponent* mesh)
{
    auto iter = std::find(mMeshComps.begin(), mMeshComps.end(), mesh);
    mMeshComps.erase(iter);
}

Texture* Renderer::GetTexture(const std::string& fileName)
{
    Texture* tex = nullptr;
    
    // Is the texture already in the map?
    auto iter = mTextures.find(fileName);
    if (iter != mTextures.end())
    {
        tex = iter->second;
    }
    else
    {
        // Load from file
        tex = new Texture();
        if (tex->Load(fileName))
        {
            mTextures.emplace(fileName, tex);
        }
        else
        {
            delete tex;
            tex = nullptr;
        }
    }
    
    return tex;
}

Mesh *Renderer::GetMesh(const std::string &fileName)
{
    Mesh *m = nullptr;
    auto iter = mMeshes.find(fileName);
    if (iter != mMeshes.end())
    {
        m = iter->second;
    }
    else
    {
        m = new Mesh();
        if (m->Load(fileName, this))
        {
            mMeshes.emplace(fileName, m);
        }
        else
        {
            delete m;
            m = nullptr;
        }
    }
    return m;
}

bool Renderer::LoadShaders()
{
    // Create sprite shader
    mSpriteShader = new Shader();
    
    // Set the view-projection matrix
    if (!mSpriteShader->Load("Shaders/Sprite.vert", "Shaders/Sprite.frag"))
    {
        return false;
    }
    
    mSpriteShader->SetActive();
    
    // Set the view-projection matrix
    Matrix4 viewProj = Matrix4::CreateSimpleViewProj(mScreenWidth, mScreenHeight);
    mSpriteShader->SetMatrixUniform("uViewProj", viewProj);
    
    // Create basic mesh shader
    mMeshShader = new Shader();
    if (!mMeshShader->Load("Shaders/Pong.vert", "Shaders/Pong.frag"))
    {
        return false;
    }

    mMeshShader->SetActive();
    
    // Set the view-projection matrix
    mView = Matrix4::CreateLookAt(
        Vector3::Zero,      // Camera position
        Vector3::UnitX,     // target position
        Vector3::UnitZ);    // Up
    mProjection = Matrix4::CreatePerspectiveFOV(
        Math::ToRadians(70.0f), // Horizontal FOV
        mScreenWidth,           // Width of view
        mScreenHeight,          // Height of view
        25.0f,                  // Near plane
        10000.0f);              // Far plane
    mMeshShader->SetMatrixUniform("uViewProj", mView * mProjection);
    
    return true;
}

void Renderer::CreateSpriteVerts()
{
    float vertices[] = {
        -0.5f,  0.5f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, // top left
         0.5f,  0.5f, 0.f, 0.f, 0.f, 0.f, 1.f, 0.f, // top right
         0.5f, -0.5f, 0.f, 0.f, 0.f, 0.f, 1.f, 1.f, // bottom right
        -0.5f, -0.5f, 0.f, 0.f, 0.f, 0.f, 0.f, 1.f  // bottom left
    };

    unsigned int indices[] = {
        0, 1, 2,
        2, 3, 0
    };

    mSpriteVerts = new VertexArray(vertices, 4, indices, 6);
}

void Renderer::SetLightUniforms(class Shader *shader)
{
    // Camera position is from inverted view
    Matrix4 invView = mView;
    invView.Invert();
    shader->SetVectorUniform("uCameraPos", invView.GetTranslation());
    
    // Ambient light
    shader->SetVectorUniform("uAmbientLight", mAmbientLight);
    
    // Directional light
    shader->SetVectorUniform("uDirLight.mDirection", mDirLight.mDirection);
    shader->SetVectorUniform("uDirLight.mDiffuseColor", mDirLight.mDiffuseColor);
    shader->SetVectorUniform("uDirLight.mSpecColor", mDirLight.mSpecColor);
}

File: Chapter06_Veldrid/Renderer.cs
using System;
using System.Collections.Generic;
using System.Numerics;
using Veldrid;
using Veldrid.Sdl2;
using Veldrid.StartupUtilities;

namespace Chapter06
{
    public class Renderer :IDisposable
    {
        private readonly Game _game;

        // Map of textures loaded
        private readonly Dictionary<string, ProcessedTexture> _textures = new();

        // Map of meshes loaded
        private readonly Dictionary<string, Mesh> _meshes = new();

        // All the sprite components drawn
        private readonly List<SpriteComponent> _spriteComponents = new();
        
        // All mesh components drawn
        private readonly List<MeshComponent> _meshComponents = new();

        // View/projection for 3D shaders
        private Matrix4x4 _view;
        private Matrix4x4 _projection;

        public Renderer(Game game)
        {
            _game = game;
        }

        public Sdl2Window Window { get; private set; }

        public GraphicsDevice GraphicsDevice { get; private set; }

        public CommandList CommandList { get; private set; }

        public VertexArray SpriteVertices { get; private set; }

        public SpriteShader SpriteShader { get; private set; }

        public MeshShader MeshShader { get; private set; }

        public bool Initialize(int screenWidth, int screenHeight)
        {
            // Create an SDL Window
            var windowCi = new WindowCreateInfo
            {
                X = 100,
                Y = 100,
                WindowWidth = screenWidth,
                WindowHeight = screenHeight,
                WindowTitle = "Game Programming in C# (Chapter 5)"
            };

            Window = VeldridStartup.CreateWindow(ref windowCi);
            if (!Window.Exists)
            {
                Console.WriteLine("Failed to create window");
                return false;
            }
            
            // Create the graphic device
            var options = new GraphicsDeviceOptions(
                debug: true,
                swapchainDepthFormat: PixelFormat.R16_UNorm,
                syncToVerticalBlank: true,
                resourceBindingModel: ResourceBindingModel.Improved,
                preferDepthRangeZeroToOne: true,
                preferStandardClipSpaceYDirection: true);
            GraphicsDevice = VeldridStartup.CreateGraphicsDevice(Window, options, GraphicsBackend.OpenGL);
            
            // Create command list
            CommandList = GraphicsDevice.ResourceFactory.CreateCommandList();
            
            // Make sure we can create/compile shaders
            if (!LoadShaders())
            {
                Console.WriteLine("Failed to load shaders.");
                return false;
            }

            // Create quad for drawing sprites
            CreateSpriteVertices();

            return true;
        }

        public void Draw()
        {
            CommandList.Begin();
            CommandList.SetFramebuffer(GraphicsDevice.SwapchainFramebuffer);
            CommandList.SetFullViewports();

            // Clear the color buffer
            CommandList.ClearColorTarget(0, new RgbaFloat(0.0f, 0.0f, 0.0f, 1.0f));

            // Draw mesh components
            CommandList.SetPipeline(MeshShader.Pipeline);
            CommandList.SetGraphicsResourceSet(0, MeshShader.ProjectionViewResourceSet);
            CommandList.UpdateBuffer(MeshShader.ProjectionViewBuffer, 0, _projection * _view);

            foreach (var mesh in _meshComponents)
            {
                mesh.Draw(MeshShader);
            }

            // Draw all sprite components
            CommandList.SetPipeline(SpriteShader.Pipeline);
            CommandList.SetGraphicsResourceSet(0, SpriteShader.ProjectionViewResourceSet);
            foreach (var sprite in _spriteComponents)
            {
                sprite.Draw(SpriteShader);
            }

            CommandList.End();
            GraphicsDevice.SubmitCommands(CommandList);

            // Once commands have been submitted, the rendered image can be presented to the application window.
            GraphicsDevice.SwapBuffers();
        }
        
        public void AddSprite(SpriteComponent sprite)
        {
            // We add it already ordered to the list
            var index = 0;
            for (; index < _spriteComponents.Count; index++)
            {
                if (sprite.DrawOrder < _spriteComponents[index].DrawOrder)
                {
                    break;
                }
            }
            
            _spriteComponents.Insert(index, sprite);
        }

        public void RemoveSprite(SpriteComponent sprite)
        {
            _spriteComponents.Remove(sprite);
        }
        
        public void AddMeshComponent(MeshComponent mesh)
        {
            _meshComponents.Add(mesh);
        }

        public void RemoveMeshComponent(MeshComponent mesh)
        {
            _meshComponents.Remove(mesh);
        }

        public ProcessedTexture GetTexture(string fileName)
        {
            if (_textures.TryGetValue(fileName, out var texture))
            {
                return texture;
            }

            texture = new ProcessedTexture();
            if (texture.Load(GraphicsDevice, fileName))
            {
                _textures.Add(fileName, texture);
            }
            else
            {
                texture.Dispose();
                texture = null;
            }

            return texture;
        }
        
        public Mesh GetMesh(string fileName)
        {
            if (_meshes.TryGetValue(fileName, out var mesh))
            {
                return mesh;
            }

            mesh = new Mesh();
            if (mesh.Load(fileName, this))
            {
                _meshes.Add(fileName, mesh);
            }
            else
            {
                mesh.Dispose();
                mesh = null;
            }

            return mesh;
        }

        public void UnloadData()
        {
            // Destroy textures
            foreach (var texture in _textures.Values)
            {
                texture.Dispose();
            }
            _textures.Clear();

            // Destroy meshes
            foreach (var mesh in _meshes.Values)
            {
                mesh.Dispose();
            }
            _meshes.Clear();
        }

        public void Dispose()
        {
            _textures.Clear();

            CommandList?.Dispose();
            SpriteVertices?.Dispose();
            SpriteShader?.Dispose();
            GraphicsDevice?.Dispose();
        }
        
        private void CreateSpriteVertices()
        {
            var vertices = new[]
            {
                new Vertex(new Vector3(-0.5f, 0.5f, 0.0f), new Vector3(0.0f, 0.0f, 0.0f), new Vector2(0.0f, 0.0f)), // top left
                new Vertex(new Vector3(0.5f, 0.5f, 0.0f), new Vector3(0.0f, 0.0f, 0.0f), new Vector2(1.0f, 0.0f)), // top right
                new Vertex(new Vector3(0.5f,-0.5f, 0.0f), new Vector3(0.0f, 0.0f, 0.0f), new Vector2(1.0f, 1.0f)), // bottom right
                new Vertex(new Vector3(-0.5f,-0.5f, 0.0f), new Vector3(0.0f, 0.0f, 0.0f), new Vector2(0.0f, 1.0f)), // bottom left
            };

            var indices = new ushort[]
            {
                0, 1, 2,
                2, 3, 0
            };

            SpriteVertices = new VertexArray(GraphicsDevice, vertices, indices);
        }

        private bool LoadShaders()
        {
            // Create Sprite Shader and Pipeline
            SpriteShader = new SpriteShader();
            if (!SpriteShader.Load(GraphicsDevice, "Shaders/Sprite.vert", "Shaders/Sprite.frag"))
            {
                return false;
            }

            // Set the view-projection matrix
            CommandList.Begin();
            Matrix4x4 viewProj = MathUtils.CreateSimpleViewProj(Window.Width, Window.Height);
            // Matrix4x4 viewProj = Matrix4x4.CreateOrthographic(Window.Width, Window.Height, -1.0f, 2.0f);
            CommandList.UpdateBuffer(SpriteShader.ProjectionViewBuffer, 0, viewProj);
            CommandList.End();

            // Create Mesh Shader and Pipeline
            MeshShader = new MeshShader();
            if (!MeshShader.Load(GraphicsDevice, "Shaders/BasicMesh.vert", "Shaders/BasicMesh.frag"))
            {
                return false;
            }

            // Set the view-projection matrix
            CommandList.Begin();
            _view = Matrix4x4.CreateLookAt(Vector3.Zero, Vector3.UnitX, Vector3.UnitZ);
            _projection = Matrix4x4.CreatePerspectiveFieldOfView(MathUtils.ToRadians(70.0f), (float)Window.Width / Window.Height, 25.0f, 10000.0f);
            CommandList.UpdateBuffer(MeshShader.ProjectionViewBuffer, 0, _projection * _view);
            CommandList.End();

            GraphicsDevice.SubmitCommands(CommandList);

            return true;
        }
    }
}